### PR TITLE
Added Test for XXE

### PR DESCRIPTION
**Pull request by DotDotSlashRepo**
This PR covers issue #8 

- [x] This PR handles the issue and requires no additional PRs. Refer TODOs
- [x] You have validated the need for this change.

**What did this PR accomplish?**

- Added test case for XXE

#### TODO
1. Edit top level page to link to test case.
2. This test case has overlap with WSTG-INPV-07 , XML injection. WSTG-INPV-07 is to be modified to eliminate duplication.
3. Need to re-order test case, XXE makes more sense to be after XML injection

Thank you for your contribution!

### CODE REVIEW

File: document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/20-Testing_for_XXE_Injection.md [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+<!ENTITY a2 "&a1;&a1;&a1;&a1;&a1;&a1;&a1;&a1;&a1;&a1;">
+<!ENTITY a3 "&a2;&a2;&a2;&a2;&a2;&a2;&a2;&a2;&a2;&a2;">
+<!ENTITY a4 "&a3;&a3;&a3;&a3;&a3;&a3;&a3;&a3;&a3;&a3;">
+]>
+<data>&a4;</data>
+```
+
+#### Filter Bypass
+
+Web application filters can be configured to detect common XXE attack payloads in HTTP request and reject such requests. Weak or poorly configured filters can be evaded using below techniques:
```
Comment (kingthorin) [inline document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/20-Testing_for_XXE_Injection.md:155]:
```suggestion
Web application filters can be configured to detect common XXE attack payloads in HTTP requests and reject such requests. Weak or poorly configured filters can be evaded using techniques outlined below:
```

File: document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/20-Testing_for_XXE_Injection.md [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+- Some XML implementations support `PUBLIC` keyword which is similar to `SYSTEM` and can be used to create external entities.
+- UTF-8 and UTF-16 encodings can be used to encode the XXE payload to get around web filters.
+- For PHP applications, PHP filters such as `convert.base64-encode` can be used to evade filters , especially if the filter logic is to identify and filter URIs starting with `file:///`.
```
Comment (kingthorin) [inline document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/20-Testing_for_XXE_Injection.md:159]:
```suggestion
- Some XML implementations support the `PUBLIC` keyword which is similar to `SYSTEM` and can be used to create external entities.
- UTF-8 and UTF-16 encodings can be used to encode the XXE payloads to get around web filters.
- For PHP applications, PHP filters such as `convert.base64-encode` can be used to evade filters, especially if the filter logic is to identify and filter URIs starting with `file:///`.
```

File: document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/20-Testing_for_XXE_Injection.md [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+### Source Code Review
+
+The following Java API may be vulnerable to XXE if they are not configured properly.
```
Comment (kingthorin) [inline document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/20-Testing_for_XXE_Injection.md:163]:
```suggestion
The following Java APIs may be vulnerable to XXE if they are not configured properly.
```

**Review by kingthorin — CHANGES_REQUESTED** [review]

**Comment by DotDotSlashRepo** [comment]
TODO

Edit top level page to link to test case.
This test case has overlap with WSTG-INPV-07 , XML injection. WSTG-INPV-07 is to be modified to eliminate duplication.
Need to re-order test case, XXE makes more sense to be after XML injection

@kingthorin : Can I go ahead with these ?

**Comment by kingthorin** [comment]
I'm good with ypu addressing duplicate content. 
However re-ordering things should wait until we are headed to 5.x. Currently we are still planning further 4.x releases, so maybe just create an issue and we can assign it to the 5.x milestone.

**Comment by kingthorin** [comment]
@DotDotSlashRepo do you intend to finish this?

**Comment by kingthorin** [comment]
@DotDotSlashRepo do you intend to finish this?

File: document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/20-Testing_for_XXE_Injection.md [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+In addition, the Java POI office reader may be vulnerable to XXE if the version is under 3.10.1.
+
+The version of POI library can be identified from the filename of the JAR. For example,
+
+- `poi-3.8.jar`
+- `poi-ooxml-3.8.jar`
+
+The followings source code keyword may apply to C.
+
+- libxml2: xmlCtxtReadMemory,xmlCtxtUseOptions,xmlParseInNodeContext,xmlReadDoc,xmlReadFd,xmlReadFile ,xmlReadIO,xmlReadMemory, xmlCtxtReadDoc ,xmlCtxtReadFd,xmlCtxtReadFile,xmlCtxtReadIO
```
Comment (rejahrehim) [inline document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/20-Testing_for_XXE_Injection.md:198]: - libxml2: `xmlCtxtReadMemory`, `xmlCtxtUseOptions`, `xmlParseInNodeContext`, `xmlReadDoc`, `xmlReadFd`, `xmlReadFile`, `xmlReadIO`, `xmlReadMemory`, `xmlCtxtReadDoc`, `xmlCtxtReadFd`, `xmlCtxtReadFile`, `xmlCtxtReadIO`

File: document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/20-Testing_for_XXE_Injection.md [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+- libxerces-c: XercesDOMParser, SAXParser, SAX2XMLReader
```
Comment (rejahrehim) [inline document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/20-Testing_for_XXE_Injection.md:199]: - libxerces-c: `XercesDOMParser`, `SAXParser`, `SAX2XMLReader`

**Comment by ThunderSon** [comment]
@RiieCco might you have some time to give this a look?

@kingthorin what is still needed for this? I'll be giving this a look as well.

**Comment by kingthorin** [comment]
Comments from Feb 5th and Rejah’s more recent comments.

**Comment by ThunderSon** [comment]
Alright. Let's re-review that comment, as that seems to be creating an issue, and is not a blocker right now.
Let's try to recall the exact concern and then move this.
This is a good addition, I'll try to review this ASAP.

**Comment by kingthorin** [comment]
I was this discussion:

> TODO
> Edit top level page to link to test case.
> This test case has overlap with WSTG-INPV-07 , XML injection. WSTG-INPV-07 is to be modified to eliminate duplication.
> Need to re-order test case, XXE makes more sense to be after XML injection

---

> I'm good with you addressing duplicate content.
> However re-ordering things should wait until we are headed to 5.x. Currently we are still planning further 4.x releases, so maybe just create an issue and we can assign it to the 5.0 milestone.

**Pull request closed without merging**